### PR TITLE
Make Augmentation Utilities More Customizable(reupload due to branch issues)

### PR DESCRIPTION
**Pull request by dewball345**
https://github.com/tensorflow/similarity/pull/249 has the details about this

### CODE REVIEW

**Comment by owenvallis** [comment]
@dewball345 looks like there are some conflicts with this branch and the vicreg module. I'll look through the other files in the meantime.

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[[source]]
```
Comment (owenvallis) [inline Pipfile:1]: I think we should add the Pipfile type to the .gitignore and remove these.

File: Pipfile.lock [hunk 2]
```diff
@@ -0,0 +1,1130 @@
+{
```
Comment (owenvallis) [inline Pipfile.lock:1]: Same here.

File: setup.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def get_version(rel_path):
     license="Apache License 2.0",
     install_requires=[
         "distinctipy",
-        "nmslib",
+        # "nmslib",
```
Comment (owenvallis) [inline setup.py:61]: Is this commented out because it won't build on your computer? I don't think we want to change this until we update the storage layer.

File: tensorflow_similarity/augmenters/augmentation_utils/cropping.py [hunk 4]
```diff
@@ -181,26 +181,80 @@ def crop_and_resize(
...
+  max_cropsize_multiplier: int = 1
+):
+  rand_width = tf.random.uniform(
+      shape=[],
+      minval=int(min_cropsize_multiplier * width),
+      maxval=int(max_cropsize_multiplier * width),
+      dtype=tf.int32,
+  )
+ 
+  # rand_height = tf.random.uniform(
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/augmentation_utils/cropping.py:199]: We should remove this if we're not using it.

File: tensorflow_similarity/augmenters/augmentation_utils/cropping.py [hunk 5]
```diff
@@ -181,26 +181,80 @@ def crop_and_resize(
...
+  #     shape=[],
+  #     minval=int(min_cropsize_multiplier * height),
+  #     maxval=int(max_cropsize_multiplier * height),
+  #     dtype=tf.int32,
+  # )
+ 
+  crop = tf.image.random_crop(image, (rand_width, rand_width, 3))
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/augmentation_utils/cropping.py:206]: We should add a doc string explaining that we always take a square crop but resize to height x width. This may stretch and distort the cropped image.

File: tensorflow_similarity/augmenters/barlow.py [hunk 6]
```diff
@@ -4,38 +4,100 @@
...
+    hue_multiplier=0.2,
+    jitter_probability=0.8,
+    greyscale_probability=0.2,
+    blur_probability=0.2,
+    blur_min_sigma=0,
+    blur_max_sigma=1,
+    solarize_probability=0.2,
+    solarize_thresh=10
+):
+    print("HI")
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/barlow.py:35]: please remove.

File: tensorflow_similarity/augmenters/barlow.py [hunk 7]
```diff
@@ -4,38 +4,100 @@
...
+                 jitter_probability=0.8,
+                 greyscale_probability=0.2,
+                 blur_probability=0.2,
+                 blur_min_sigma=0,
+                 blur_max_sigma=1,
+                 solarize_probability=0.2,
+                 solarize_thresh=10,
                  num_cpu: Optional[int] = os.cpu_count(),
          ):
+        # print("HI")
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/barlow.py:84]: Same here.

File: tensorflow_similarity/augmenters/barlow.py [hunk 8]
```diff
@@ -45,21 +107,41 @@ def augment(
...
-
+ 
         with tf.device("/cpu:0"):
             inputs = tf.stack(x)
-            inputs = tf.cast(inputs, dtype="float32") / 255.0
+            inputs = tf.cast(inputs, dtype="float32") / 255
             views = []
-
+ 
+            #NOTE: do i need to normalize here? or before
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/barlow.py:116]: This is tricky as most images should be uint8 255, but that's not always true. If the user already called an image transform function, then we definitely wouldn't want to normalize here. I'd likely avoid this step as I don't think any of the tf.image functions require the images to be in a specific range.

File: tensorflow_similarity/losses/vicreg.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
+from typing import Any, Callable, Dict, Optional
```
Comment (owenvallis) [inline tensorflow_similarity/losses/vicreg.py:1]: Vicreg is already in development and this version seems to cause a conflict. Should we remove this from this PR?

**Review by owenvallis — CHANGES_REQUESTED** [review]
Thanks for the PR. Looks good, left some comments and looks like we can likely remove the vicreg module as it's already in the dev branch.

File: tensorflow_similarity/losses/vicreg.py (hunk 9 above)
Comment (dewball345) [reply tensorflow_similarity/losses/vicreg.py:1]: I believe i changed some important things - previous version had some small issues regarding tf ops and customizability. This should replace previous vicreg - have confirmed this version works pretty well.

**Comment by dewball345** [comment]
@owenvallis thanks for the review, i'll quickly address these changes end of today or tomorrow

File: tensorflow_similarity/augmenters/barlow.py (hunk 8 above)
Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:116]: I believe solarization does require normalization - i think its a good idea to ask if the user wants normalization via a parameter in the constructor and apply it accordingly.

Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:116]: Something that I had noticed was that when normalizing in the augmenter fn(ex. augment_barlow), the model has huge losses in performance which confuses me.

File: tensorflow_similarity/augmenters/augmentation_utils/cropping.py (hunk 5 above)
Comment (dewball345) [reply tensorflow_similarity/augmenters/augmentation_utils/cropping.py:206]: Okay, i think i've added it

File: tensorflow_similarity/augmenters/barlow.py (hunk 8 above)
Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py]: BTW, something that I had noticed was that when normalizing in the augmenter fn(ex. augment_barlow), the model has huge losses in performance which confuses me.

Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py]: Okay, i've made normalization optional so that it can be enabled/disabled. Also, have changed augmentation_utils.solarize so that it can accept non-normalized images.

File: setup.py (hunk 3 above)
Comment (dewball345) [reply setup.py]: Oops, my bad! That was supposed to be uncommented. NMSLib doesn't build on windows for python version >= 3.9, so I uncommented it and re-ran everything on a Python 3.8 kernel. Commenting it was for debugging purposes and is fixed now.

**Comment by dewball345** [comment]
Okay, I think i have completed or resolved the changes.

File: tensorflow_similarity/losses/vicreg.py (hunk 9 above)
Comment (owenvallis) [reply tensorflow_similarity/losses/vicreg.py]: Sounds good. I went ahead and resolved the merge conflicts by taking all the changes from this PR.

**Comment by dewball345** [comment]
@owenvallis Okay, looks like all tests have passed

File: tensorflow_similarity/augmenters/barlow.py [hunk 10]
```diff
@@ -30,9 +31,8 @@ def augment_barlow(
     blur_min_sigma=0,
     blur_max_sigma=1,
     solarize_probability=0.2,
-    solarize_thresh=10
+    solarize_thresh=10,
```
Comment (owenvallis) [inline tensorflow_similarity/augmenters/barlow.py:34]: This is the only problematic parameter, as the value depends on the range of the image.  To make solarize work, I think we need the user to pass us the expected min and max image values and a threshold.  Requiring the user to pass this avoids the need for us to guess at what the correct range should be.

Something like the following should work for solarized for all value ranges ([0,1], [0, 255], or even [-127.5, 127.5]):

def solarize(img, min, max, thresh):
  midpoint = (max - min)/2.
  adjusted_img = img - min - midpoint
  adjusted_thresh = thresh - min - midpoint
  img = np.where(adjusted_img < adjusted_thresh, adjusted_img, -adjusted_img)
  img = img + min + midpoint
  return img

wdyt?

**Review by owenvallis — CHANGES_REQUESTED** [review]

File: tensorflow_similarity/augmenters/barlow.py (hunk 10 above)
Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:34]: sounds good. Will try to make these changes tomorrow

Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:33]: Why don't we do something like:

```py
def solarize(img, min, max, thresh):
     img = np.where(img < thresh, img, max - img)
```

instead? why do we have to adjust the images?

File: tensorflow_similarity/augmenters/barlow.py [hunk 11]
```diff
@@ -19,6 +19,7 @@ def augment_barlow(
     image: tf.Tensor, 
     height: int, 
     width: int,
+    normalize_img=True,
```
[the remark this answers is not in the document]
Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py]: Okay, i've removed the normalize parameter

File: tensorflow_similarity/augmenters/barlow.py (hunk 10 above)
Comment (owenvallis) [reply tensorflow_similarity/augmenters/barlow.py:35]: The simpler version is cleaner but I think it only supports positive values. If a client passes in an image with min = -127.5, max = 127.5 and thresh = -117.5, then we'll end up mapping -117.5 to (125.5 - -117.5) = 243. 

I suppose it doesn't matter though if the other transforms only support positive values as well.

Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:35]: Okay, good to know. BTW, have noticed that the tests failed, but not quite sure why.

Comment (owenvallis) [reply tensorflow_similarity/augmenters/barlow.py:35]: I've come across this before. Looks like the TF version updated and there were some changes to the tf.keras.application models. Seems like they added an extra layer. We can just update the expected number in the test, and I should make those tests less brittle at some point.

Comment (owenvallis) [reply tensorflow_similarity/augmenters/barlow.py:35]: Let's just merge it and I can fix them up on my end. 👍

Comment (dewball345) [reply tensorflow_similarity/augmenters/barlow.py:35]: Okay sure, you can merge this PR then :)

**Review by owenvallis — APPROVED** [review]
